File: client/src/shared/lib/e2e/ratchet.ts
import { chacha20poly1305 } from '@noble/ciphers/chacha.js'
import { hkdf } from '@noble/hashes/hkdf.js'
import { hmac } from '@noble/hashes/hmac.js'
import { sha256 } from '@noble/hashes/sha2.js'

import { concatBytes, fromBase64, toBase64, toUtf8 } from './codec'
import { diffieHellman, generateKeyPair, type KeyPair } from './keys'

/**
 * The Double Ratchet, ported from `server/pkg/e2e/ratchet.go`.
 *
 * Forward secrecy (a leaked key does not open past messages) plus
 * post-compromise security (a DH ratchet step heals the session). The server is
 * a blind relay throughout — it moves ciphertext and a header and can decrypt
 * neither.
 *
 * The header is authenticated, not encrypted: it travels as additional data for
 * the AEAD, so tampering with the ratchet key or counters fails decryption.
 */

export interface RatchetHeader {
  /** sender's current ratchet public key */
  dh: Uint8Array
  /** number of messages in the previous sending chain */
  pn: number
  /** message number in the current sending chain */
  n: number
}

export class DecryptError extends Error {
  constructor() {
    super('e2e: decryption failed')
    this.name = 'DecryptError'
  }
}

/** Bounds the work a hostile header can force by claiming a huge gap. */
const MAX_SKIP = 1000

/**
 * Serialises a header exactly as Go's `json.Marshal` does.
 *
 * This is the AEAD's additional data, so the bytes must match the Go side
 * character for character. Two details carry the whole compatibility:
 * Go renders a `[]byte` field as standard base64, and it emits struct fields in
 * declaration order (dh, pn, n) with no whitespace.
 */
export function marshalHeader(header: RatchetHeader): Uint8Array {
  return toUtf8(JSON.stringify({ dh: toBase64(header.dh), pn: header.pn, n: header.n }))
}

export function unmarshalHeader(bytes: Uint8Array): RatchetHeader {
  const parsed = JSON.parse(new TextDecoder().decode(bytes)) as {
    dh: string
    pn: number
    n: number
  }
  return { dh: fromBase64(parsed.dh), pn: parsed.pn ?? 0, n: parsed.n ?? 0 }
}

/** Derives (rootKey, chainKey) from the root key and a DH output. */
function kdfRootKey(rootKey: Uint8Array, dhOut: Uint8Array): [Uint8Array, Uint8Array] {
  const out = hkdf(sha256, dhOut, rootKey, toUtf8('Synapse-Ratchet-RK'), 64)
  return [out.slice(0, 32), out.slice(32)]
}

/** Advances a chain key; the constants 0x02/0x01 are part of the wire contract. */
function kdfChainKey(chainKey: Uint8Array): [Uint8Array, Uint8Array] {
  const nextChainKey = hmac(sha256, chainKey, new Uint8Array([0x02]))
  const messageKey = hmac(sha256, chainKey, new Uint8Array([0x01]))
  return [nextChainKey, messageKey]
}

/**
 * Derives the AEAD key and nonce from a message key.
 *
 * A fixed nonce is safe here precisely because the key is derived per message
 * and never reused — the message key itself is deliberately not used as the
 * cipher key.
 */
function aeadFor(messageKey: Uint8Array): { key: Uint8Array; nonce: Uint8Array } {
  const buf = hkdf(sha256, messageKey, undefined, toUtf8('Synapse-Ratchet-Msg'), 32 + 12)
  return { key: buf.slice(0, 32), nonce: buf.slice(32) }
}

function seal(messageKey: Uint8Array, ad: Uint8Array, plaintext: Uint8Array): Uint8Array {
  const { key, nonce } = aeadFor(messageKey)
  return chacha20poly1305(key, nonce, ad).encrypt(plaintext)
}

function open(messageKey: Uint8Array, ad: Uint8Array, ciphertext: Uint8Array): Uint8Array {
  const { key, nonce } = aeadFor(messageKey)
  return chacha20poly1305(key, nonce, ad).decrypt(ciphertext)
}

/** Storage key for a skipped message key. Internal only — never crosses the wire. */
function skippedKey(dhPublic: Uint8Array, n: number): string {
  return `${toBase64(dhPublic)}|${n}`
}

/** A session serialised for persistence across reloads. */
export interface SerializedSession {
  dhsPrivate: string
  dhsPublic: string
  dhr: string | null
  rootKey: string
  sendingChainKey: string | null
  receivingChainKey: string | null
  sent: number
  received: number
  previousSent: number
  skipped: Record<string, string>
}

/**
 * One Double Ratchet session with one peer device.
 *
 * Not safe for concurrent use — callers serialise per session, exactly as the
 * Go implementation requires.
 */
export class RatchetSession {
  private dhs: KeyPair
  private dhr: Uint8Array | null
  private rootKey: Uint8Array
  private sendingChainKey: Uint8Array | null = null
  private receivingChainKey: Uint8Array | null = null
  private sent = 0
  private received = 0
  private previousSent = 0
  private skipped = new Map<string, Uint8Array>()

  private constructor(dhs: KeyPair, dhr: Uint8Array | null, rootKey: Uint8Array) {
    this.dhs = dhs
    this.dhr = dhr
    this.rootKey = rootKey
  }

  /**
   * Initiator's session. The peer's signed prekey is the first DHr, and one DH
   * ratchet runs immediately so the initiator can send straight away.
   */
  static initiator(sharedSecret: Uint8Array, theirSignedPreKey: Uint8Array): RatchetSession {
    const session = new RatchetSession(generateKeyPair(), theirSignedPreKey, sharedSecret)
    const dhOut = diffieHellman(session.dhs.privateKey, theirSignedPreKey)
    const [rootKey, chainKey] = kdfRootKey(session.rootKey, dhOut)
    session.rootKey = rootKey
    session.sendingChainKey = chainKey
    return session
  }

  /**
   * Responder's session. It has no sending chain until the first message
   * arrives and triggers a ratchet step.
   */
  static responder(sharedSecret: Uint8Array, signedPreKey: KeyPair): RatchetSession {
    return new RatchetSession(signedPreKey, null, sharedSecret)
  }

  encrypt(plaintext: Uint8Array): { header: RatchetHeader; ciphertext: Uint8Array } {
    if (!this.sendingChainKey) throw new Error('e2e: no sending chain')
    const [nextChainKey, messageKey] = kdfChainKey(this.sendingChainKey)
    this.sendingChainKey = nextChainKey

    const header: RatchetHeader = {
      dh: this.dhs.publicKey,
      pn: this.previousSent,
      n: this.sent,
    }
    this.sent++
    return { header, ciphertext: seal(messageKey, marshalHeader(header), plaintext) }
  }

  decrypt(header: RatchetHeader, ciphertext: Uint8Array): Uint8Array {
    // 1. A key stored for a message that arrived out of order.
    const stored = this.trySkipped(header, ciphertext)
    if (stored) return stored

    // 2. A new ratchet key means the peer stepped; follow them.
    if (!this.sameDhr(header.dh)) {
      this.skipMessageKeys(header.pn)
      this.dhRatchet(header)
    }

    // 3. Catch up on anything missing earlier in this chain.
    this.skipMessageKeys(header.n)

    // 4. Derive this message's key and open it.
    if (!this.receivingChainKey) throw new DecryptError()
    const [nextChainKey, messageKey] = kdfChainKey(this.receivingChainKey)
    this.receivingChainKey = nextChainKey
    this.received++
    try {
      return open(messageKey, marshalHeader(header), ciphertext)
    } catch {
      throw new DecryptError()
    }
  }

  private sameDhr(dhPublic: Uint8Array): boolean {
    if (!this.dhr || this.dhr.length !== dhPublic.length) return false
    return this.dhr.every((byte, index) => byte === dhPublic[index])
  }

  /** Advances to the peer's new ratchet key: new receiving chain, then a new sending chain. */
  private dhRatchet(header: RatchetHeader): void {
    this.previousSent = this.sent
    this.sent = 0
    this.received = 0
    this.dhr = header.dh

    const [rootKey, receivingChainKey] = kdfRootKey(
      this.rootKey,
      diffieHellman(this.dhs.privateKey, this.dhr),
    )
    this.rootKey = rootKey
    this.receivingChainKey = receivingChainKey

    this.dhs = generateKeyPair()
    const [nextRootKey, sendingChainKey] = kdfRootKey(
      this.rootKey,
      diffieHellman(this.dhs.privateKey, this.dhr),
    )
    this.rootKey = nextRootKey
    this.sendingChainKey = sendingChainKey
  }

  /** Stores keys for messages we have not seen yet, so a late arrival still opens. */
  private skipMessageKeys(until: number): void {
    if (!this.receivingChainKey || !this.dhr) return
    if (until - this.received > MAX_SKIP) {
      throw new Error('e2e: too many skipped messages')
    }
    while (this.received < until) {
      const [nextChainKey, messageKey] = kdfChainKey(this.receivingChainKey)
      this.receivingChainKey = nextChainKey
      this.skipped.set(skippedKey(this.dhr, this.received), messageKey)
      this.received++
    }
  }

  private trySkipped(header: RatchetHeader, ciphertext: Uint8Array): Uint8Array | null {
    const key = skippedKey(header.dh, header.n)
    const messageKey = this.skipped.get(key)
    if (!messageKey) return null
    try {
      const plaintext = open(messageKey, marshalHeader(header), ciphertext)
      // A message key is single-use; keeping it would allow a replay to succeed.
      this.skipped.delete(key)
      return plaintext
    } catch {
      return null
    }
  }

  serialize(): SerializedSession {
    return {
      dhsPrivate: toBase64(this.dhs.privateKey),
      dhsPublic: toBase64(this.dhs.publicKey),
      dhr: this.dhr ? toBase64(this.dhr) : null,
      rootKey: toBase64(this.rootKey),
      sendingChainKey: this.sendingChainKey ? toBase64(this.sendingChainKey) : null,
      receivingChainKey: this.receivingChainKey ? toBase64(this.receivingChainKey) : null,
      sent: this.sent,
      received: this.received,
      previousSent: this.previousSent,
      skipped: Object.fromEntries(
        [...this.skipped].map(([key, value]) => [key, toBase64(value)]),
      ),
    }
  }

  static deserialize(state: SerializedSession): RatchetSession {
    const session = new RatchetSession(
      { privateKey: fromBase64(state.dhsPrivate), publicKey: fromBase64(state.dhsPublic) },
      state.dhr ? fromBase64(state.dhr) : null,
      fromBase64(state.rootKey),
    )
    session.sendingChainKey = state.sendingChainKey ? fromBase64(state.sendingChainKey) : null
    session.receivingChainKey = state.receivingChainKey
      ? fromBase64(state.receivingChainKey)
      : null
    session.sent = state.sent
    session.received = state.received
    session.previousSent = state.previousSent
    session.skipped = new Map(
      Object.entries(state.skipped ?? {}).map(([key, value]) => [key, fromBase64(value)]),
    )
    return session
  }
}

export { concatBytes }

File: server/pkg/e2e/ratchet.go
package e2e

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"errors"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
)

// NewInitiatorSession starts Alice's session after X3DH. sk is the shared secret;
// theirSignedPreKey is Bob's signed prekey public (Alice's initial DHr).
func NewInitiatorSession(sk, theirSignedPreKey []byte) (*Session, error) {
	dhr, err := PublicKeyFromBytes(theirSignedPreKey)
	if err != nil {
		return nil, err
	}
	dhs, err := GenerateKeyPair()
	if err != nil {
		return nil, err
	}
	s := &Session{dhs: dhs, dhr: dhr, rk: sk, skipped: map[string][]byte{}}
	// Perform the initial DH ratchet so Alice has a sending chain.
	dhOut, err := dh(s.dhs.Priv, s.dhr)
	if err != nil {
		return nil, err
	}
	s.rk, s.cks = kdfRK(s.rk, dhOut)
	return s, nil
}

// NewResponderSession starts Bob's session. sk is the shared secret; signedPreKey
// is Bob's signed prekey key pair (his initial DHs). Bob has no sending chain
// until he receives Alice's first message and ratchets.
func NewResponderSession(sk []byte, signedPreKey *KeyPair) (*Session, error) {
	return &Session{dhs: signedPreKey, rk: sk, skipped: map[string][]byte{}}, nil
}

// Encrypt ratchet-encrypts plaintext, returning the header and ciphertext.
func (s *Session) Encrypt(plaintext []byte) (Header, []byte, error) {
	var mk []byte
	s.cks, mk = kdfCK(s.cks)
	hdr := Header{DH: s.dhs.PublicBytes(), PN: s.pn, N: s.ns}
	s.ns++
	ad := hdr.bytes()
	ct, err := aeadSeal(mk, ad, plaintext)
	if err != nil {
		return Header{}, nil, err
	}
	return hdr, ct, nil
}

// Decrypt ratchet-decrypts a message, performing a DH ratchet step and/or
// skipped-key handling as needed.
func (s *Session) Decrypt(hdr Header, ciphertext []byte) ([]byte, error) {
	// 1. If we have a stored skipped key for this header, use it.
	if pt, ok := s.trySkipped(hdr, ciphertext); ok {
		return pt, nil
	}
	// 2. If the header advertises a new ratchet key, perform a DH ratchet.
	if !s.sameDHr(hdr.DH) {
		if err := s.skipMessageKeys(hdr.PN); err != nil {
			return nil, err
		}
		if err := s.dhRatchet(hdr); err != nil {
			return nil, err
		}
	}
	// 3. Skip any messages before this one in the current receiving chain.
	if err := s.skipMessageKeys(hdr.N); err != nil {
		return nil, err
	}
	// 4. Derive this message's key and decrypt.
	var mk []byte
	s.ckr, mk = kdfCK(s.ckr)
	s.nr++
	pt, err := aeadOpen(mk, hdr.bytes(), ciphertext)
	if err != nil {
		return nil, ErrDecrypt
	}
	return pt, nil
}

func (s *Session) sameDHr(dhPub []byte) bool {
	return s.dhr != nil && string(s.dhr.Bytes()) == string(dhPub)
}

// dhRatchet advances to the peer's new ratchet key: derive a new receiving
// chain, then a new sending chain with a fresh local ratchet key.
func (s *Session) dhRatchet(hdr Header) error {
	newDHr, err := PublicKeyFromBytes(hdr.DH)
	if err != nil {
		return err
	}
	s.pn = s.ns
	s.ns = 0
	s.nr = 0
	s.dhr = newDHr

	dhOut, err := dh(s.dhs.Priv, s.dhr)
	if err != nil {
		return err
	}
	s.rk, s.ckr = kdfRK(s.rk, dhOut)

	// Generate a new local ratchet key pair and derive the new sending chain.
	s.dhs, err = GenerateKeyPair()
	if err != nil {
		return err
	}
	dhOut2, err := dh(s.dhs.Priv, s.dhr)
	if err != nil {
		return err
	}
	s.rk, s.cks = kdfRK(s.rk, dhOut2)
	return nil
}

// skipMessageKeys derives and stores message keys up to (not including) until,
// so out-of-order / missing messages can still be decrypted later.
func (s *Session) skipMessageKeys(until uint32) error {
	if s.ckr == nil {
		return nil
	}
	if until-s.nr > uint32(maxSkip) {
		return errors.New("e2e: too many skipped messages")
	}
	for s.nr < until {
		var mk []byte
		s.ckr, mk = kdfCK(s.ckr)
		s.skipped[skKey(s.dhr.Bytes(), s.nr)] = mk
		s.nr++
	}
	return nil
}

func (s *Session) trySkipped(hdr Header, ct []byte) ([]byte, bool) {
	key := skKey(hdr.DH, hdr.N)
	mk, ok := s.skipped[key]
	if !ok {
		return nil, false
	}
	pt, err := aeadOpen(mk, hdr.bytes(), ct)
	if err != nil {
		return nil, false
	}
	delete(s.skipped, key)
	return pt, true
}

func skKey(dhPub []byte, n uint32) string {
	var nb [4]byte
	binary.BigEndian.PutUint32(nb[:], n)
	return string(dhPub) + "|" + string(nb[:])
}

// --- KDFs ---

// kdfRK derives (newRootKey, chainKey) from the root key and a DH output.
func kdfRK(rk, dhOut []byte) (newRK, ck []byte) {
	out := make([]byte, 64)
	r := hkdf.New(sha256.New, dhOut, rk, []byte("Synapse-Ratchet-RK"))
	_, _ = io.ReadFull(r, out)
	return out[:32], out[32:]
}

// kdfCK advances a chain key: HMAC with constant inputs gives (newCK, messageKey).
func kdfCK(ck []byte) (newCK, mk []byte) {
	mac := hmac.New(sha256.New, ck)
	mac.Write([]byte{0x02})
	newCK = mac.Sum(nil)
	mac.Reset()
	mac.Write([]byte{0x01})
	mk = mac.Sum(nil)
	return newCK, mk
}

// --- AEAD (ChaCha20-Poly1305) ---

func aeadSeal(mk, ad, plaintext []byte) ([]byte, error) {
	aead, nonce, err := aeadFor(mk)
	if err != nil {
		return nil, err
	}
	// Deterministic zero nonce is safe: each message key is unique (used once).
	return aead.Seal(nonce[:0], nonce, plaintext, ad), nil
}

func aeadOpen(mk, ad, ciphertext []byte) ([]byte, error) {
	aead, nonce, err := aeadFor(mk)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, ciphertext, ad)
}

func aeadFor(mk []byte) (aead interface {
	Seal(dst, nonce, plaintext, additionalData []byte) []byte
	Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error)
}, nonce []byte, err error) {
	// Derive an AEAD key + nonce from the message key via HKDF so the message key
	// itself is never used directly as the cipher key.
	buf := make([]byte, chacha20poly1305.KeySize+chacha20poly1305.NonceSize)
	r := hkdf.New(sha256.New, mk, nil, []byte("Synapse-Ratchet-Msg"))
	if _, err = io.ReadFull(r, buf); err != nil {
		return nil, nil, err
	}
	c, err := chacha20poly1305.New(buf[:chacha20poly1305.KeySize])
	if err != nil {
		return nil, nil, err
	}
	return c, buf[chacha20poly1305.KeySize:], nil
}

func (h Header) bytes() []byte {
	b, _ := json.Marshal(h)
	return b
}

// MarshalHeader / UnmarshalHeader serialize a header for the wire.
func MarshalHeader(h Header) []byte { return h.bytes() }

// UnmarshalHeader parses a header from the wire.
func UnmarshalHeader(b []byte) (Header, error) {
	var h Header
	err := json.Unmarshal(b, &h)
	return h, err
}
